presto-matching/src/main/java/com/facebook/presto/matching/Match.java

```diff
@@ -13,191 +13,68 @@
  */
 package com.facebook.presto.matching;
 
-import java.util.NoSuchElementException;
-import java.util.function.Function;
-import java.util.function.Predicate;
+import java.util.Objects;
 
+import static com.google.common.base.MoreObjects.toStringHelper;
 import static java.util.Objects.requireNonNull;
 
-public abstract class Match<T>
+public final class Match<T>
 {
-    public abstract boolean isPresent();
-
-    public abstract T value();
-
-    public final boolean isEmpty()
+    public static <T> Match<T> of(T value, Captures captures)
     {
-        return !isPresent();
+        return new Match<>(value, captures);
     }
 
-    public static <S> Match<S> of(S value, Captures captures)
-    {
-        Captures result;
-        requireNonNull(captures);
-        return new Match.Present<>(value, captures);
-    }
+    private final T value;
```

Comment:
maybe we just remove `T value` from match? It seems that's not needed by optimizer (it knows which node is matched). This would simplify the code. We can always store matched objects in captures.

Reply:
You'll loose typesafety if you do that. Also, that would simplify the library code, but would complicate the user-site code (the optimizer)

Reply:
It would also be useful if we every implement `find/findAnywhare` type of pattern lookup, but it makes the code mode complex right now.